gpu/containers/include/pcl/gpu/containers/device_memory.h

```diff
@@ -36,222 +36,262 @@
+
+  /** \brief Copy constructor. Just increments reference counter. */
+  DeviceMemory2D(const DeviceMemory2D& other_arg);
+
+  /** \brief Assignment operator. Just increments reference counter. */
+  DeviceMemory2D&
+  operator=(const DeviceMemory2D& other_arg);
+
+  /** \brief Allocates internal buffer in GPU memory. If internal buffer was created
+   * before the function recreates it with new size. If new and old sizes are equal it
+   * does nothing.
+   * \param rows_arg number of rows to allocate
+   * \param colsBytes_arg width of the buffer in bytes
+   * */
+  void
+  create(int rows_arg, int colsBytes_arg);
+
+  /** \brief Decrements reference counter and releases internal buffer if needed. */
+  void
+  release();
+
+  /** \brief Performs data copying. If destination size differs it will be reallocated.
+   * \param other destination container
+   * */
+  void
+  copyTo(DeviceMemory2D& other) const;
+
+  /** \brief Uploads data to internal buffer in GPU memory. It calls create() inside to
+   * ensure that intenal buffer size is enough.
+   * \param host_ptr_arg pointer to host buffer to upload
+   * \param host_step_arg stride between two consecutive rows in bytes for host buffer
+   * \param rows_arg number of rows to upload
+   * \param colsBytes_arg width of host buffer in bytes
+   * */
+  void
+  upload(const void* host_ptr_arg,
+         std::size_t host_step_arg,
+         int rows_arg,
+         int colsBytes_arg);
+
+  /** \brief Downloads data from internal buffer to CPU memory. User is responsible for
+   * correct host buffer size.
+   * \param host_ptr_arg pointer to host buffer to download
+   * \param host_step_arg stride between two consecutive rows in bytes for host buffer
+   * */
+  void
+  download(void* host_ptr_arg, std::size_t host_step_arg) const;
+
+  /** \brief Performs swap of data pointed with another device memory.
+   * \param other_arg device memory to swap with
+   * */
+  void
+  swap(DeviceMemory2D& other_arg);
+
+  /** \brief Returns pointer to given row in internal buffer.
+   * \param y_arg row index
+   * */
+  template <class T>
+  T*
+  ptr(int y_arg = 0);
+
+  /** \brief Returns constant pointer to given row in internal buffer.
+   * \param y_arg row index
+   * */
+  template <class T>
+  const T*
+  ptr(int y_arg = 0) const;
+
+  /** \brief Conversion to PtrStep for passing to kernel functions. */
+  template <class U>
+  operator PtrStep<U>() const;
+
+  /** \brief Conversion to PtrStepSz for passing to kernel functions. */
+  template <class U>
+  operator PtrStepSz<U>() const;
+
+  /** \brief Returns true if unallocated otherwise false. */
+  bool
+  empty() const;
+
+  /** \brief Returns number of bytes in each row. */
+  int
+  colsBytes() const;
+
+  /** \brief Returns number of rows. */
+  int
+  rows() const;
+
+  /** \brief Returns stride between two consecutive rows in bytes for internal buffer.
+   * Step is stored always and everywhere in bytes!!! */
```

Comment:
I think the line regarding `Step is stored... bytes!!!` can be converted to a note or a warning. Not relevant to this PR